app.go

```diff
@@ -471,14 +471,19 @@ func (app *app) loop() {
 	}
 }
 
+type cleanFunc func()
+
+var noopCleanFunc = func() {}
+
 // This function is used to run a shell command. Modes are as follows:
 //
 //	Prefix  Wait  Async  Stdin  Stdout  Stderr  UI action
 //	$       No    No     Yes    Yes     Yes     Pause and then resume
 //	%       No    No     Yes    Yes     Yes     Statline for input/output
 //	!       Yes   No     Yes    Yes     Yes     Pause and then resume
 //	&       No    Yes    No     No      No      Do nothing
-func (app *app) runShell(s string, args []string, prefix string) {
+//	$|      No    No     Pipe   Yes     Yes     Pause, Pipe execute, return cleanup to resume
+func (app *app) runShell(s string, args []string, prefix string) cleanFunc {
```

Comment:
Instead of returning a cleanup function, I would either make `$|` a separate function or (this is probably easier):

- rename this function to `runShellInternal` and add an `sendToStdin string` argument to it.
- create a new `runShell` function that calls `runShellInternal` and passes `""` for `sendToStdin`
- create a `runShellWithStding` function with arguments `s`, `args`, and `sendToStdin` that calls `runShellInternal(s, args, "$|", sendToStdin)`.

Perhaps we can think of something more elegant than exactly what I suggested. However, we should avoid returning functions that the caller must to remember to call in some cases, but are completely ignored in other cases.

Reply:
That's certainly a good idea! I did thought of the possibility that `cleanup` is not called. Although I figured the docs should help.

Separating the `runShell` in different modes should certainly help with cleanness of the code. Although existing `runShell` feels a bit cramped up as well, there is a good separation of start, run/start, cleanup state in it, but the switch cases had to be duplicated a couple of times. I think `$`, `%` etc should be separated as well.

Anyway, that's just some of my thoughts. I'll definitely try to get this in another function as the `$|` is a internal one. I have the `list-mappings` etc in a branch ready to go. I'll modify all this in the other PR. sending it soon.